New neural OT solver ENOT

nazarblch opened the pull request:
This pull request includes a new method (solver) for Neural OT training, called ENOT. It solves the dual OT problem in d-dimensional Euclidean spaces with the use of specific expectile regularization on the Kantorovich potentials.

The method is similar to W2NeuralDual (already implemented in the repository) but extends its functionality with the ability to use different cost functions, and eliminates the need for computationally intensive fine-tuning with conjugate optimizers.

Our paper arXiv:2403.03777 provides a detailed description of the method and its experimental evaluation.  ENOT outperforms previous state-of-the-art approaches on the Wasserstein-2 benchmark tasks by a large margin (up to a 3-fold improvement in quality and up to a 10-fold improvement in runtime). 

We have also included in this pull request a detailed tutorial, demonstrating how our method works. Please find the source code with ENOT implementation and the tutorial with applications on some 2D datasets by the following links:
https://github.com/nazarblch/ott/blob/dev/src/ott/neural/solvers/expectile_neural_dual.py
https://github.com/nazarblch/ott/blob/dev/docs/tutorials/ENOT.ipynb

michalk8 commented:
Hi @nazarblch , thanks a lot for the contribution! Will give it a review in the upcoming days!

marcocuturi commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    return self._grad_f(source)
+
+
+class EuclideanPotentials(DualPotentials):
```
I think we have a varriable `corr` in the original potentials class to enable this slightly different logic, would it work here? (at least the `transport` method would not need to be changed)

nazarblch replied:
I agree, overriding methods was unnecessary in this class and we may use `corr` instead. I just suggested that it would be better to use inheritance instead of `corr`  in `DualPotentials`.

nazarblch replied:
Class EuclideanPotentials removed.

marcocuturi commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+
+
+@jax.tree_util.register_pytree_node_class
+class DotCost(costs.CostFn):
```
if this is needed, maybe better add it in `costs.py`

nazarblch replied:
If it's useful somewhere other than this solver, I can move it to costs.py.

nazarblch replied:
Class DotCost removed. Replaced by `lambda x, y: -jax.vmap(jnp.dot)(x, y)`

marcocuturi commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  r"""Expectile-regularized Neural Optimal Transport (ENOT) :cite:`buzun:24`.
+
+  It solves the dual optimal transport problem for a specified cost function
+  c(x, y) between two measures :math:`\alpha` and :math:`\beta` in
```
here the cost needs to be `TICost`?

nazarblch replied:
The cost function doesn't have to be TICost. We use TICost only in combination with is_bidirectional option and there is a corresponding assert:
`assert (
        isinstance(cost_fn, (costs.TICost, DotCost)) or not is_bidirectional
    ), (
        "is_bidirectional=True can only be used with a translation invariant"
        "cost (TICost) or DotCost."
    )`

marcocuturi reviewed (commented):
Thanks a lot @nazarblch !! This is just a few very basic comments. I will check the paper + code, and Michal will also start a review

michalk8 commented:
Hi @nazarblch , I started looking into this PR, could you please adapt to the new structure as introduced in introduced #468 ?

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  Implements a potential value or a vector field.
+  """
+
+  is_potential: bool  # Implements a potential value or a vector field.
```
Please add these comments as docstrings.

nazarblch replied:
Moved to the docstrings.

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  add_l2_norm: bool  # If true, l2 norm is added to the potential.
+  model: nn.Module  # The potential model.
+
+  def apply(
```
Please don't overwrite `apply`, rather the `__call__` method.

nazarblch replied:
Rewritten using `__call__` method.

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+                                  )
+
+
+class PotentialModelWrapper(nn.Module):
```
Might be worthwhile subclassing `ott.neural.networks.potentials.BasePotential` and overriding `potential_value_fn` instead.

michalk8 replied:
In doing so, we get `potential_gradient_fn` for free, as well as `create_train_state`.

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    )
+
+
+class MLP(nn.Module):
```
Am thinking of the following:
with the wrapper in mind, we could move this to `ott.neural.networks.potentials` (including the wrapper) and remove the `PotentialMLP` (both of the 2 classes you define are more general) it seems.
Pinging @marcocuturi @Algue-Rythme for opinion.

nazarblch replied:
I have left this class in the same place for compatibility, but support the idea to replace `PotentialMLP` in the future.

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    return nn.Dense(self.dim_hidden[-1])(x)
+
+
+class ExpectileNeuralDual(Generic[CostFn_t]):
```
Why the use of generics? Imho not necessary, would remove.

nazarblch replied:
Generic removed.

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    logging: option to return logs.
+    rng: random key used for seeding for network initializations.
+    expectile: parameter of the expectile loss (:math:`\tau`).
+      Suggested values range is [0.9, 1.0).
```
Let's use :math:\`[0.9, 1.0)\`.

nazarblch replied:
rewritten as proposed

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    expectile: parameter of the expectile loss (:math:`\tau`).
+      Suggested values range is [0.9, 1.0).
+    expectile_loss_coef: expectile loss weight.
+      Suggested values range is [0.3, 1.0].
```
Similarly here, :math:\`[0.3, 1.0]\`.

nazarblch replied:
rewritten as proposed

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    self.is_bidirectional = is_bidirectional
+    self.use_dot_product = (type(cost_fn) == DotCost)
+
+    assert (
```
Would rewrite as
```python
if is_birectional:
   assert isinstance(cost_fn, (costs.TICost, DotCost)
```

nazarblch replied:
rewritten as suggested

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    self.expectile = expectile
+    self.expectile_loss_coef = expectile_loss_coef
+    self.is_bidirectional = is_bidirectional
+    self.use_dot_product = (type(cost_fn) == DotCost)
```
Use `type(cost_fn) is DotCost`.

nazarblch replied:
DotCost removed

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+                        PotentialValueFn_t], batch: Dict[str, jnp.ndarray]
+  ) -> Tuple[jnp.ndarray, Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray]]:
+
+    if self.is_bidirectional:
```
Wouldn't it be possible to do instead:
```python
transport = self.to_dual_potentials().transport
```

michalk8 replied:
If not, please refactor using the twist operator.

nazarblch replied:
Replaced by `transport = ENOTPotentials(
        gradient_f(params_f), g_value(params_g), self.cost_fn,
        self.is_bidirectional, self.use_dot_product
    ).transport`

michalk8 commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+
+    return loss, (dual_loss, amor_loss, w_dist)
+
+  def _euclidean_loss_fn(
```
Possibly a naive question, but why is `DotCost` treated differently here - shouldn't it be equivalent (at least in theory the `def _loss_fn` with the `DotCost`? Or is this more numerically stable impl.?

michalk8 reviewed (changes requested):
@nazarblch thanks a lot for the contribution! Still need to finish looking at `_loss_fn` and `_euclidean_loss_fn`'s implementation + the notebook, but left other comments.

nazarblch commented:
> @nazarblch thanks a lot for the contribution! Still need to finish looking at `_loss_fn` and `_euclidean_loss_fn`'s implementation + the notebook, but left other comments.

This is a short confirmation that I have read the comments and am working on them.

nazarblch commented on docs/spelling/misc.txt:
```diff
@@ -15,13 +16,16 @@ gmm
 loc
 logsumexp
 lse
+mathbb
```
Removed

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+from flax import struct
+from flax.core import frozen_dict
+from flax.training import train_state
+from jax._src.basearray import Array as Array
```
Removed

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    if self.is_potential:
+      z = z.squeeze()
+
+    if self.is_potential and self.add_l2_norm:
```
rewritten as proposed

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    return jnp.ones((1, dim))
+
+
+class PotentialTrainState(train_state.TrainState):
```
Replaced by `ott.neural.networks.potentials.PotentialTrainState`.

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+      jnp.mean(jnp.sum(tgt ** 2, axis=-1))
+
+
+def expectile_loss(
```
I have moved `expectile_loss` into `ExpectileNeuralDual` with adding a short docstring.

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  """
+
+  def __init__(
+      self, g: PotentialValueFn_t, grad_f: PotentialGradientFn_t,
```
I have changed the order.

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    return vec - self._grad_h_inv(self._grad_g(vec))
+
+
+class ForwardPotentials(DualPotentials):
```
I have merged these three classes into one (ENOTPotentials).

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  r"""The dual potentials for only forward mapping.
+
+  Args:
+  g: The second dual potential function.
```
done

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  g: The second dual potential function.
+  grad_f: Gradient of the first dual potential function.
+  cost_fn: The translation invariant (TI) cost function used to solve
+  the OT problem.
```
done

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+
+
+class BidirectionalPotentials(DualPotentials):
+  r"""The dual potentials for bidirectional mapping with TICost.
```
r removed

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+Callback_t = Callable[[int, DualPotentials], None]
+PotentialValueFn_t = Callable[[jnp.ndarray], jnp.ndarray]
+PotentialGradientFn_t = Callable[[jnp.ndarray], jnp.ndarray]
+CostFn_t = TypeVar("CostFn_t", bound=costs.CostFn)
```
`CostFn_t` removed

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  r"""The dual potentials for bidirectional mapping with TICost.
+
+  Args:
+  g: The second dual potential function.
```
done

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
```
agree

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+    super().__init__(g_cost_conjugate, g, cost_fn=cost_fn, corr=False)
+    self.__grad_f = grad_f
+
+    assert isinstance(cost_fn, costs.TICost), (
```
I have removed assertions in potentials class because they are the same as in `ExpectileNeuralDual`.

nazarblch commented on src/ott/neural/solvers/expectile_neural_dual.py:
```diff
@@ -0,0 +1,655 @@
+
+Train_t = Dict[Literal["train_logs", "valid_logs"], Dict[str, List[float]]]
+Callback_t = Callable[[int, DualPotentials], None]
+PotentialValueFn_t = Callable[[jnp.ndarray], jnp.ndarray]
```
I have imported `PotentialValueFn_t` and `PotentialGradientFn_t` from `ott.neural.networks.potentials`.

michalk8 commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+  r"""Expectile-regularized Neural Optimal Transport (ENOT) :cite:`buzun:24`.
+
+  It solves the dual optimal transport problem for a specified cost function
+  c(x, y) between two measures :math:`\alpha` and :math:`\beta` in
```
Let's use :math:\`c(x, y)\`

nazarblch replied:
corrected

michalk8 commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+  The :class:`flax.linen.Module` potentials for ``neural_f``
+  and ``neural_g`` can
+
+  1. both provide the values of the potentials :math:`f` and :math:`g`, or
```
Please mention explicitly this is controlled by `is_bidirectional`.

nazarblch replied:
done

michalk8 commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+
+  Args:
+    dim_data: Input dimensionality of data required for network init.
+    neural_f: Network architecture for potential :math:`f`.
```
Network architecture for potential :math:`f` or its gradient :math:\`\nabla f\`.

nazarblch replied:
corrected

michalk8 commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+    return lambda x: self.apply({"params": params}, x)
+
+
+class MLP(nn.Module):
```
I'd either:
1. move this to `ott.neural.networks.potentials`
2. or remove this + the default networks in `ExpectileNeuralDual` and just add it to the notebook as a reference potential.

nazarblch replied:
Moved to `ott.neural.networks.potentials` to be consistent with `W2NeuralDual` .

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+
+import jax
+import jax.numpy as jnp
+from jax.lax import stop_gradient
```
the  import  removed

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+
+import optax
+from flax import linen as nn
+from flax.core.frozen_dict import FrozenDict
```
corrected

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+    conjugate_cost = jnp.dot if corr else cost_fn
+
+    def g_cost_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      y_hat = jax.lax.stop_gradient(transport(x))
```
I have inserted the code above. However, there doesn't seem to be much difference.

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+
+  def __init__(
+      self, grad_f: PotentialGradientFn_t, g: PotentialValueFn_t,
+      cost_fn: costs.CostFn, is_bidirectional: bool, corr: bool
```
... *  added

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport from source to the target distribution or vice-versa."""
```
I have removed the docstring.

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+
+from ott import utils
+from ott.geometry import costs
+from ott.neural.networks.potentials import (
```
Replaced by potentials import

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+    PotentialTrainState,
+    PotentialValueFn_t,
+)
+from ott.problems.linear.potentials import DualPotentials
```
corrected

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+  where :math:`\mathcal{L}_{\tau}` is the least asymmetrically weighted
+  squares loss from expectile regression.
+
+  The :class:`flax.linen.Module` potentials for ``neural_f``
```
removed

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+    vec = jnp.atleast_2d(vec)
+    assert (forward
+            is True), ("Only forward mapping (source -> target) is supported.")
+    return self._grad_f(vec)
```
No, `T(x) = grad(f)(x)` in case `self.is_bidirectional=False`  and independent on `corr`.   If `self.is_bidirectional=False`  we can use any function for the transport mapping T(x), including `x - grad(f)(x)`  (since we don't express it through the potential f). But the last transform  ``x - grad(f)(x)`  may be unexpected by user and `grad(f)(x)` seems to be more straightforward. I have also updated it in the new version of paper.

michalk8 replied:
Thank you for the explanation, makes perfect sense!

nazarblch commented on docs/neural/methods.rst:
```diff
@@ -19,6 +19,13 @@ Neural Dual
+.. autosummary::
+    :toctree: _autosummary
+
+    expectile_neural_dual.ExpectileNeuralDual
```
I agree with the proposal to move to the Neural Dual section

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+
+    return loss, (f_loss, g_loss, w_dist)
+
+  def to_dual_potentials(self) -> DualPotentials:
```
corrected

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+
+    @jax.jit
+    def step_fn(state_f, state_g, batch):
+      grad_fn = jax.value_and_grad(self._loss_fn, argnums=[0, 1], has_aux=True)
```
I agree, ` jax.value_and_grad` is not required

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+      return super().transport(vec, forward)
+
+    vec = jnp.atleast_2d(vec)
+    assert (forward
```
corrected

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+  def tree_flatten(self) -> Tuple[Sequence[Any], Dict[str, Any]]:
+    """Flatten the kwargs."""
+    return [], {
+        "grad_f": self._grad_f,
```
corrected

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+      self, grad_f: PotentialGradientFn_t, g: PotentialValueFn_t,
+      cost_fn: costs.CostFn, is_bidirectional: bool, corr: bool
+  ):
+
```
removed

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+      )
+
+    if use_dot_product:
+      assert isinstance(
```
moved above

nazarblch commented on src/ott/neural/methods/expectile_neural_dual.py:
```diff
@@ -0,0 +1,526 @@
+    )
+
+    rng = utils.default_prng_key(rng)
+    rng, rng_f, rng_g = jax.random.split(rng, 3)
```
corrected

nazarblch commented:
@michalk8, I have corrected the code and updated the repository. Also we uploaded a new version of the paper into arxiv with additional image to image translation experiments.

michalk8 reviewed (approved):
Thanks a lot for your contribution @nazarblch , I just briefly polished some of the docs, LGTM!
